### DISCUSSION TITLE
Let's Talk about it

### DISCUSSION THREAD

**Discussion by Mbucari** [Mbucari/aaxclean-cli#1] [General]
@mkb79

I added a new release.  It includes updates to AAXClean for speed improvements as well as binaries that do not require the .net runtime.  The linux64 program compiled for no-framework is 26.4MB

Also, in your email you commented that AAXClean's m4b files were slightly larger than ffmpeg's.  Mpeg4 data is stored in "samples", and samples are grouped into "chunks".  The number of samples per chunk is a choice made by the encoder.  AAXClean preserved the samples/chunk groupings present in the aax file.  FFMpeg squashes many more samples into fewer chunks. As a result, FFMpeg's chunk offset table and sample to chunk table are smaller.

**Comment by mkb79** [comment]
> I added a new release. It includes updates to AAXClean for speed improvements as well as binaries that do not require the .net runtime. The linux64 program compiled for no-framework is 26.4MB

I'll test it and give you my feedback. I'd want to replace ffmpeg with your aaxclean. Maybe you can add an option to replace the chapters with the one from the API chapter metadata via a json file?

**Comment by mkb79** [comment]
> Also, in your email you commented that AAXClean's m4b files were slightly larger than ffmpeg's. Mpeg4 data is stored in "samples", and samples are grouped into "chunks". The number of samples per chunk is a choice made by the encoder. AAXClean preserved the samples/chunk groupings present in the aax file. FFMpeg squashes many more samples into fewer chunks. As a result, FFMpeg's chunk offset table and sample to chunk table are smaller.

Thank you for these technical information. 0.05 % diff is very small. If all metadata (cover, tags, etc.) unchanged than all is fine!

Best regards and thank you for your very good work!

**Comment by Mbucari** [comment]
I added support for json chapters in v0.1.4.

Argument --chapter_info accepts a file path to a json file containing Audible's chapter_info structure (found in content_metadata).

**Reply by mkb79** [reply]
Oh, I read it late. Sorry. Thank you for your work!

**Comment by jives56** [comment]
Hi, sorry if this is like butting in but I'm interested in your project and would like to use it. My question is where does one get the audible_key and audible_iv from? There is a 41 byte companion file with my downloaded aaxc file. Thanks for your clarification.

**Reply by Mbucari** [reply]
The audible_key and audible_iv are file-specific decryption parameters.  When an audible client requests a download link for aaxc files, it also receives a "voucher" for the downloaded file, and the voucher contains those decryption parameters.

So where/how you get them depends on how you got your aaxc file.  If you got the aaxc files from [Libation](https://github.com/rmcrackan/Libation) (with the "Retain aaxc files" option), Libation saves a .key file with the decryption key and iv in plaintext.

Where did you get your aaxc files and 41-byte mystery file?

**Comment by Rumia-Channel** [comment]
Hello. I have recently come across this software, and I was wondering if it is possible to divide the data directly into chapters, as displayed by --list-chapters?

I would like to divide by the chapter metadata that the aaxc file has, but I don't know how.

Also, what are the units for the numbers in the example "Chapter 1|0|1345245"? 
Seconds? Milliseconds?

**Reply by Mbucari** [reply]
They're supposed to be milliseonds.

I added an -s option to split the file into chapters. If you don't provide chapters, it will use chapter info from metadata.
